---
layout: page
category: cook
title: "Paired reads"
order: 4
---

For SRA accessions Magic-BLAST determines whether reads are paired and maps them appropriately.

For reads in FASTA and FASTQ files paired reads can either be in a single file, or two files.

##### Single file

For paired reads presented as successive entries in a single FASTA or FASTQ
file, i.e. read 1 and 2 of fragment 1, then read 1 and 2 of fragment 2,
etc., simply add the parameter ```-paired```:

```
magicblast -query reads.fa -db genome -paired
```

or

```
magicblast -query reads.fastq -db genome -paired -infmt fastq
```

##### Two files

For paired reads presented in two parallel files, use these options:

```
magicblast -query reads.fa -query_mate mates.fa -db genome
```

or

```
magicblast -query reads.fastq -query_mate mates.fastq -db genome -infmt fastq
```
