---
layout: page
category: release
title: "Release Notes"
order: 0
---

#### Magic-BLAST 1.4.0: June 18, 2018

##### New features
* New option: ```-no_discordant``` to report only concordant read pairs
* Report strand-specific alignments with ```-fr``` and ```-rf``` flags for forward-reveresed and reversed-forward


##### Improvements
* Improved sensitivity: count for frequent database words to be filtered was increased to 60 (used to be 10). This threshold can be changed with the ```-max_db_word_count``` option.
* Non-cannonical splice signal now require longer alignments on the exon, with score at least 50 on both sides of an intron.


##### Bug fixes
* Fix for alignments stopping prematurely 


---


#### Magic-BLAST 1.3.0: September 15, 2017

##### New features
* The alignment cutoff score (```-score``` option) can be expressed as either a constant or a function of read length in this format: L,b,a for a * length + b
* Maximum edit distance cutoff for alignments can be specified with ```-max_edit_dist``` option
* SRA caching is now turned off by default and can be turned on with the ```-sra_cache``` flag. When accessing sequences in NCBI SRA repository the data can be cached in a local file so that it is not downloaded over the network again when reused later. It may result in very large local files and is only needed if you use SRA accessions multiple times, have very limited network bandwidth and a large disk space.


##### Improvements
* Unmapped reads are reported in the SAM and Tabular reports, use ```-no_unaligned``` option to not have unaligned reads reported
* The XS tag (used by transcript assemblers) is now reported in SAM output
* The score threshold (```-score``` option) now applies to the whole spliced alignment
* The query batch size (number of reads processed at a time) is now controlled with environment variable BATCH_SIZE expressed in cumulative number of bases
* The default mismatch and gap extension penalties are now set to 4
* Improved sensitivity and run time


##### Bug fixes
* Read ids for pairs are printed properly in the SAM format (no ".1" ".2" or "/1" "/2")
* The secondary alignment bit (256) is set in SAM flags
* Maximum intron length option ```-max_intron_length``` works properly


---


#### Magic-BLAST 1.2.0: February 17, 2017

##### Improvements

* Improved multi-threading for larger genomes
* Improved splice site detection

##### Bug fixes

* Magic-BLAST now works with multiple SRA accessions
* Fixed the macOS dmg installer that used to remove BLAST+ binaries
* The -seqidlist option is no longer ignored


---


#### Magic-BLAST 1.1.0: November 4, 2016

##### Improvements

* -sra option connects to NCBI via HTTPS
* Results are formatted with 'bare' accessions
* Tabular output includes a header with column titles

##### Bug fixes:

* Fixed SAM flag values


---


#### Magic-BLAST 1.0.0: August 19, 2016

* First release

